gen/src/AddGroupNoise.cxx

```diff
@@ -0,0 +1,155 @@
+  , m_rng_tn(rng)
+  , m_nsamples(4096)
+  , m_fluctuation(0.1)  // Fluctuation randomization
+  , m_shift(0.9)
+  , m_period(0.4)
+  , log(Log::logger("sim"))
+{
+}
+
+Gen::AddGroupNoise::~AddGroupNoise() {}
+
+void Gen::AddGroupNoise::gen_elec_resp_default()
+{
+    // calculate the frequencies ...
+    m_elec_resp_freq.resize(m_fft_length, 0);
+    for (unsigned int i = 0; i != m_elec_resp_freq.size(); i++) {
+        if (i <= m_elec_resp_freq.size() / 2.) {
+            m_elec_resp_freq.at(i) =
+                i / (m_elec_resp_freq.size() * 1.0) * 1. / m_period;  // the second half is useless ...
+        }
+        else {
+            m_elec_resp_freq.at(i) = (m_elec_resp_freq.size() - i) / (m_elec_resp_freq.size() * 1.0) * 1. /
+                                     m_period;  // the second half is useless ...
+        }
+    }
+}
+
+WireCell::Configuration Gen::AddGroupNoise::default_configuration() const
+{
+    Configuration cfg;
+
+    cfg["spectra_file"] = m_spectra_file;
+    cfg["map_file"] = m_map_file;
+    cfg["rng"] = m_rng_tn;
+    cfg["nsamples"] = m_nsamples;
+    cfg["random_fluctuation_amplitude"] = m_fluctuation;
+    cfg["random_shift_amplitude"] = m_shift;
+    cfg["period"] = m_period;
+    cfg["normalization"] = m_normalization;
+
+    return cfg;
+}
+
+void Gen::AddGroupNoise::configure(const WireCell::Configuration& cfg)
+{
+    m_rng_tn = get(cfg, "rng", m_rng_tn);
+    m_rng = Factory::find_tn<IRandom>(m_rng_tn);
+    m_spectra_file = get(cfg, "spectra_file", m_spectra_file);
+    m_map_file = get(cfg, "map_file", m_map_file);
+    m_nsamples = get<int>(cfg, "nsamples", m_nsamples);
+    m_period = get<int>(cfg, "period", m_period);
+    m_fluctuation = get<double>(cfg, "random_fluctuation_amplitude", m_fluctuation);
+    m_shift = get<double>(cfg,"random_shift_amplitude",m_shift);
+    m_normalization = get<int>(cfg, "normalization", m_normalization);
+
+    m_fft_length = fft_best_length(m_nsamples);
+    gen_elec_resp_default();
+    
+    m_ch2grp.clear();
+    auto mapdata = Persist::load(m_map_file);
+    for(unsigned int i=0; i<mapdata.size();++i){
+	auto jdata = mapdata[i];
+	int groupID = jdata["groupID"].asInt();
+	auto chnls_json = jdata["channels"];
+	for(unsigned int j=0;j<chnls_json.size();++j){
+		m_ch2grp.insert({chnls_json[j].asInt(),groupID});
+	} 
+    }
+    
+    
+
+    m_grp2spec.clear();
+    m_grp2rnd_amp.clear();
+    m_grp2rnd_phs.clear();
+    auto specdata = Persist::load(m_spectra_file);
+    
+    for(unsigned int i=0; i<mapdata.size();++i){
+	auto jdata = specdata[i];
+	int groupID = jdata["groupID"].asInt();
+	auto spec_freq = jdata["freqs"];
+	auto spec_amps =  jdata["amps"];
+	std::vector<float> spec(spec_freq.size());
+        std::vector<float> rnd_phs(spec_freq.size());
+        std::vector<float> rnd_amp(spec_freq.size());
+	for(unsigned int j=0;j<spec.size();++j){
+		spec[j]=spec_amps[j].asFloat();
+		rnd_amp[j] = m_shift + 2 * m_fluctuation * m_rng->uniform(0, 1);
```

Comment:
Randomness should not be applied once in `configure()` but once per group in `operator()`.  Also, sampling complex amplitude is a Rayleigh distribution (it samples a radius).